phat/contracts/codebase/lib.rs

```diff
@@ -0,0 +1,333 @@
+            self.managers.clone()
+        }
+
+        /// Allows the manager to upgrade the contract
+        /// Fails if the caller is not a manager or if the new code hash is not found.
+        #[ink(message)]
+        pub fn self_upgrade(&mut self, code_hash: Hash) -> Result<()> {
+            self.ensure_manager()?;
+            self.env()
+                .set_code_hash(&code_hash)
+                .or(Err(Error::CodeNotFound))
+        }
+
+        /// Fetches basic information about the codebase
+        #[ink(message)]
+        pub fn info(&self) -> BaseInfo {
+            self.base_info.clone()
+        }
+
+        /// Sets the maximum size allowed for a single piece of code
+        /// Can only be called by a manager.
+        #[ink(message)]
+        pub fn set_max_code_size(&mut self, max_code_size: u64) -> Result<()> {
+            self.ensure_manager()?;
+            info!("Setting max code size to: {max_code_size:?}");
+            self.base_info.max_code_size = max_code_size;
+            Ok(())
+        }
+
+        /// Sets the total capacity allowed for storing codes
+        /// Can only be called by a manager.
+        #[ink(message)]
+        pub fn set_capacity(&mut self, capacity: u64) -> Result<()> {
+            self.ensure_manager()?;
+            info!("Setting total capacity to: {capacity:?}");
+            self.base_info.capacity = capacity;
+            Ok(())
+        }
+
+        /// Allows uploading of a new piece of code
+        /// Checks for size constraints and updates metadata accordingly.
+        #[ink(message)]
+        pub fn upload(&mut self, code: String) -> Result<Hash> {
+            if code.as_bytes().len() as u64 > self.base_info.max_code_size {
+                error!("Uploaded code exceeds maximum size limit.");
+                return Err(Error::CodeTooLarge);
+            }
+            let code_hash: Hash = self.env().hash_bytes::<Blake2x256>(code.as_bytes()).into();
+            if self.codes.contains(code_hash) {
+                info!("Code already exists. Returning existing hash.");
+                return Ok(code_hash);
+            }
+            if self.metadatas.contains(code_hash) {
+                info!("Metadata already exists for the code. Returning hash.");
+                return Ok(code_hash);
+            }
+            if self.base_info.total_size.saturating_add(code.len() as u64) > self.base_info.capacity
+            {
+                error!("Uploading code exceeds total capacity.");
+                return Err(Error::CapacityExceeded);
+            }
+            let metadata = CodeMetadata {
+                ref_cnt: 0,
+                uploaded_at_block: Self::env().block_number(),
+            };
+            self.metadatas.insert(code_hash, &metadata);
+            self.codes.insert(code_hash, &code);
+            self.env().emit_event(CodeUploaded {
+                code_hash,
+                uploader: Self::env().caller(),
+            });
```

Comment:
Move this after base_info updates since there can be potential overflow in `self.base_info.count`.

Reply:
1. the `self.base_info.count` is of type `u64` which is impossible to overflow in practice.
2. Events is part of the transaction, which means it can only be observed when the entire function success. If something fails after the event emit, it wouldn't appear in the log or event chain.